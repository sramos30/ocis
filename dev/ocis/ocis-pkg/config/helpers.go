package config

import (
	"io/fs"
	"os"
	"path/filepath"
	"strings"

	gofig "github.com/gookit/config/v2"
	gooyaml "github.com/gookit/config/v2/yaml"
)

var (
	defaultLocations = []string{
		filepath.Join(os.Getenv("HOME"), "/.ocis/config/"),
		"/etc/ocis/",
		".config/",
	}

	// supportedExtensions is determined by gookit/config.
	supportedExtensions = []string{
		"yaml",
		"yml",
	}
)

// DefaultConfigSources returns a slice with matched expected config files. It sugars coat several aspects of config file
// management by assuming there are 3 default locations a config file could be.
// It uses globbing to match a config file by name, and retrieve any supported extension supported by our drivers.
// It sanitizes the output depending on the list of drivers provided.
func DefaultConfigSources(filename string, drivers []string) []string {
	var sources []string

	locations := []string{}
	if v := os.Getenv("OCIS_CONFIG_DIR"); v != "" {
	    locations = append(locations, v)
		// only use the configured config dir
		locations = append(locations, os.Getenv("OCIS_CONFIG_DIR"))
	} else {
		// merge config from all default locations
		locations = append(locations, defaultLocations...)
	}

	for i := range locations {
		dirFS := os.DirFS(locations[i])
		pattern := filename + ".*"
		matched, _ := fs.Glob(dirFS, pattern)
		if len(matched) > 0 {
			// prepend path to results
			for j := 0; j < len(matched); j++ {
				matched[j] = filepath.Join(locations[i], matched[j])
			}
		}
		sources = append(sources, matched...)
	}

	return sanitizeExtensions(sources, drivers, func(a, b string) bool {
		return strings.HasSuffix(filepath.Base(a), b)
	})
}

// sanitizeExtensions removes elements from "set" which extensions are not in "ext".
func sanitizeExtensions(set []string, ext []string, f func(a, b string) bool) []string {
	var r []string
	for i := 0; i < len(set); i++ {
		for j := 0; j < len(ext); j++ {
			if f(filepath.Base(set[i]), ext[j]) {
				r = append(r, set[i])
			}
		}
	}
	return r
}

// BindSourcesToStructs assigns any config value from a config file / env variable to struct `dst`. Its only purpose
// is to solely modify `dst`, not dealing with the config structs; and do so in a thread safe manner.
func BindSourcesToStructs(extension string, dst interface{}) (*gofig.Config, error) {
	sources := DefaultConfigSources(extension, supportedExtensions)
	cnf := gofig.NewWithOptions(extension)
	cnf.WithOptions(func(options *gofig.Options) {
		options.DecoderConfig.TagName = "yaml"
	})
	cnf.AddDriver(gooyaml.Driver)
	_ = cnf.LoadFiles(sources...)

	err := cnf.BindStruct("", &dst)
	if err != nil {
		return nil, err
	}

	return cnf, nil
}
