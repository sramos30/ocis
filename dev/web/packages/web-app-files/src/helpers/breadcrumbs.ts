import { bus } from 'web-pkg/src/instance'

export interface BreadcrumbItem {
  text: string
  to?: string
  allowContextActions?: boolean
  onClick?: () => void
}

export const breadcrumbsFromPath = (
  currentPath: string,
  resourcePath: string
): BreadcrumbItem[] => {
  const pathSplit = (p = '') => p.split('/').filter(Boolean)
  const current = pathSplit(currentPath)
  const resource = pathSplit(resourcePath)

  return resource.map(
    (text, i) =>
      ({
        allowContextActions: true,
        text,
        to: '/' + [...current].splice(0, current.length - resource.length + i + 1).join('/')
      } as BreadcrumbItem)
  )
}

export const concatBreadcrumbs = (...items: BreadcrumbItem[]): BreadcrumbItem[] => {
  const last = items.pop()

  return [
    ...items,
    {
      allowContextActions: last.allowContextActions,
      text: last.text,
      onClick: () => bus.publish('app.files.list.load')
    }
  ]
}
