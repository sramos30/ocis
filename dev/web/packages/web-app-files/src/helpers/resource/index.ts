export * from './asset'
export * from './filter'
export * from './privatePreviewBlob'
export * from './publicPreviewUrl'
export * from './resource'
export * from './sameResource'
