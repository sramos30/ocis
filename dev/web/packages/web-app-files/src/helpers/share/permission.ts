// dummy to trick gettext string extraction into recognizing strings
const $gettext = (str) => {
  return str
}

export class SharePermission {
  private readonly _key: string
  private readonly _bit: number
  private readonly _label: string

  constructor(key: string, bit: number, label: string) {
    this._key = key
    this._bit = bit
    this._label = label
  }

  get key(): string {
    return this._key
  }

  get bit(): number {
    return this._bit
  }

  get label(): string {
    return $gettext(this._label)
  }

  public enabled(bitmask: number): boolean {
    return (bitmask & this.bit) > 0
  }
}

export abstract class SharePermissions {
  static readonly read = new SharePermission('read', 1, $gettext('Read'))
  static readonly update = new SharePermission('update', 2, $gettext('Update'))
  static readonly create = new SharePermission('create', 4, $gettext('Create'))
  static readonly delete = new SharePermission('delete', 8, $gettext('Delete'))
  static readonly share = new SharePermission('share', 16, $gettext('Share'))

  static permissionsToBitmask(permissions: SharePermission[]): number {
    return (permissions || []).reduce((b: number, p: SharePermission) => b | p.bit, 0)
  }

  static bitmaskToPermissions(bitmask: number): SharePermission[] {
    return [this.read, this.update, this.create, this.delete, this.share].reduce(
      (a: SharePermission[], p: SharePermission) => {
        if (p.enabled(bitmask)) {
          a.push(p)
        }
        return a
      },
      []
    )
  }
}
