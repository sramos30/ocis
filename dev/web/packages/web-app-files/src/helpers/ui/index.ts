export * as fileList from './filesList'
export * as resourceTable from './resourceTable'
