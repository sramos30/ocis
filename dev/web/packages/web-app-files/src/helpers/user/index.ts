export * from './avatarUrl'
export * from './types'
