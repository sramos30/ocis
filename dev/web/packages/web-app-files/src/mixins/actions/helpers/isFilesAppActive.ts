import { activeApp } from 'web-pkg/src/composables'

const isFilesAppActive = (activeApp: string): boolean => {
  // FIXME: we should use this constant but it somehow breaks the unit tests
  // return activeApp === FilesApp.appInfo.id
  return activeApp === 'files'
}

export default {
  computed: {
    $_isFilesAppActive(): boolean {
      return isFilesAppActive(activeApp(this.$route))
    }
  }
}
