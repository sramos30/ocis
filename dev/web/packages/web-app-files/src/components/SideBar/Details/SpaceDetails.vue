<template>
  <div id="oc-space-details-sidebar">
    <div class="oc-space-details-sidebar-image oc-text-center">
      <oc-spinner v-if="loadImageTask.isRunning" />
      <div v-else-if="spaceImage" class="oc-position-relative">
        <img :src="spaceImage" alt="" class="oc-mb-m" />
      </div>
      <oc-icon
        v-else
        name="layout-grid"
        size="xxlarge"
        class="space-default-image oc-px-m oc-py-m"
      />
    </div>
    <div
      v-if="hasMemberShares || hasLinkShares"
      class="oc-flex oc-flex-middle oc-mb-m oc-text-small"
    >
      <div v-if="hasMemberShares" class="oc-flex oc-flex-middle">
        <oc-button
          appearance="raw"
          :aria-label="$gettext('Open the member panel')"
          @click="expandMemberPanel"
        >
          <oc-icon name="group" class="oc-mr-s" />
        </oc-button>
      </div>
      <div v-if="hasLinkShares" class="oc-flex oc-flex-middle">
        <oc-button
          appearance="raw"
          :aria-label="$gettext('Open the link panel')"
          @click="expandLinkPanel"
        >
          <oc-icon name="link" class="oc-mr-s" />
        </oc-button>
      </div>
      <div>
        <span v-if="hasMemberShares">
          <span v-text="memberShareLabel" />
          <oc-button
            appearance="raw"
            :aria-label="$gettext('Open the member panel')"
            size="small"
            @click="expandMemberPanel"
          >
            <span class="oc-text-small" v-text="$gettext('Show')"></span>
          </oc-button>
        </span>
        <span v-if="hasMemberShares && hasLinkShares">Â·</span>
        <span v-if="hasLinkShares">
          <span v-text="linkShareLabel" />
          <oc-button
            appearance="raw"
            :aria-label="$gettext('Open the link panel')"
            size="small"
            @click="expandLinkPanel"
          >
            <span class="oc-text-small" v-text="$gettext('Show')"></span>
          </oc-button>
        </span>
      </div>
    </div>
    <div>
      <table class="details-table" :aria-label="detailsTableLabel">
        <tr>
          <th scope="col" class="oc-pr-s" v-text="$gettext('Last activity')" />
          <td v-text="lastModifyDate" />
        </tr>
        <tr v-if="space.description">
          <th scope="col" class="oc-pr-s" v-text="$gettext('Subtitle')" />
          <td v-text="space.description" />
        </tr>
        <tr>
          <th scope="col" class="oc-pr-s" v-text="$gettext('Manager')" />
          <td>
            <span v-text="ownerUsernames" />
          </td>
        </tr>
        <tr>
          <th scope="col" class="oc-pr-s" v-text="$gettext('Quota')" />
          <td>
            <space-quota :space-quota="space.spaceQuota" />
          </td>
        </tr>
      </table>
    </div>
  </div>
</template>
<script lang="ts">
import { defineComponent, ref } from '@vue/composition-api'
import Mixins from '../../../mixins'
import MixinResources from '../../../mixins/resources'
import { mapActions, mapGetters } from 'vuex'
import { useTask } from 'vue-concurrency'
import { buildResource, buildWebDavSpacesPath } from '../../../helpers/resources'
import { spaceRoleManager } from '../../../helpers/share'
import SpaceQuota from '../../SpaceQuota.vue'
import { loadPreview } from '../../../helpers/resource'
import { ImageDimension } from '../../../constants'

export default defineComponent({
  name: 'SpaceDetails',
  components: { SpaceQuota },
  mixins: [Mixins, MixinResources],
  inject: ['displayedItem'],
  setup() {
    const spaceImage = ref('')

    const loadImageTask = useTask(function* (signal, ref) {
      if (!ref.space?.spaceImageData) {
        return
      }

      const webDavPathComponents = ref.space.spaceImageData.webDavUrl.split('/')
      const path = webDavPathComponents
        .slice(webDavPathComponents.indexOf(ref.space.id) + 1)
        .join('/')

      const fileInfo = yield ref.$client.files.fileInfo(buildWebDavSpacesPath(ref.space.id, path))
      const resource = buildResource(fileInfo)

      spaceImage.value = yield loadPreview({
        resource,
        isPublic: false,
        dimensions: ImageDimension.Preview,
        server: ref.configuration.server,
        userId: ref.user.id,
        token: ref.getToken
      })
    })

    return { loadImageTask, spaceImage }
  },
  computed: {
    ...mapGetters('Files', [
      'highlightedFile',
      'currentFileOutgoingCollaborators',
      'currentFileOutgoingLinks'
    ]),
    ...mapGetters(['user', 'getToken']),

    space() {
      return this.displayedItem.value
    },
    detailsTableLabel() {
      return this.$gettext('Overview of the information about the selected space')
    },
    lastModifyDate() {
      return this.formDateFromISO(this.space.mdate)
    },
    ownerUsernames() {
      const userId = this.user?.id
      return this.currentFileOutgoingCollaborators
        .filter((share) => share.role.name === spaceRoleManager.name)
        .map((share) => {
          if (share.collaborator.name === userId) {
            return this.$gettextInterpolate(this.$gettext('%{displayName} (me)'), {
              displayName: share.collaborator.displayName
            })
          }
          return share.collaborator.displayName
        })
        .join(', ')
    },
    hasMemberShares() {
      return this.memberShareCount > 1
    },
    hasLinkShares() {
      return this.linkShareCount > 0
    },
    memberShareCount() {
      return this.currentFileOutgoingCollaborators.length
    },
    linkShareCount() {
      return this.currentFileOutgoingLinks.length
    },
    memberShareLabel() {
      return this.$gettextInterpolate(
        this.$ngettext(
          'This space has %{memberShareCount} member.',
          'This space has %{memberShareCount} members.',
          this.memberShareCount
        ),
        {
          memberShareCount: this.memberShareCount
        }
      )
    },
    linkShareLabel() {
      return this.$gettextInterpolate(
        this.$ngettext(
          '%{linkShareCount} link giving access.',
          '%{linkShareCount} links giving access.',
          this.linkShareCount
        ),
        {
          linkShareCount: this.linkShareCount
        }
      )
    }
  },
  watch: {
    'space.spaceImageData': {
      handler(val) {
        if (!val) {
          return
        }
        this.loadImageTask.perform(this)
      },
      deep: true
    }
  },
  mounted() {
    this.loadImageTask.perform(this)
  },
  methods: {
    ...mapActions('Files/sidebar', {
      setSidebarPanel: 'setActivePanel',
      closeSidebar: 'close'
    }),

    expandMemberPanel() {
      this.setSidebarPanel('space-share-item')
    },

    expandLinkPanel() {
      this.setSidebarPanel('links-item')
    }
  }
})
</script>
<style lang="scss" scoped>
.oc-space-details-sidebar {
  &-image img {
    max-height: 150px;
    object-fit: cover;
    width: 100%;
  }
}

.details-table {
  text-align: left;

  tr {
    height: 1.5rem;
  }

  th {
    font-weight: 600;
  }
}
</style>
