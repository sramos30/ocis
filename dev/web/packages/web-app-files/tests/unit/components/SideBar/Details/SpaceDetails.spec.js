import { createLocalVue, shallowMount } from '@vue/test-utils'
import Vuex from 'vuex'
import SpaceDetails from '../../../../../src/components/SideBar/Details/SpaceDetails.vue'
import stubs from '../../../../../../../tests/unit/stubs'
import GetTextPlugin from 'vue-gettext'
import AsyncComputed from 'vue-async-computed'
import VueCompositionAPI from '@vue/composition-api/dist/vue-composition-api'
import { ShareTypes, spaceRoleManager } from '../../../../../src/helpers/share'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(AsyncComputed)
localVue.use(VueCompositionAPI)
localVue.use(GetTextPlugin, {
  translations: 'does-not-matter.json',
  silent: true
})
const OcTooltip = jest.fn()

const spaceMock = {
  type: 'space',
  name: ' space',
  id: '1',
  mdate: 'Wed, 21 Oct 2015 07:28:00 GMT',
  spaceQuota: {
    used: 100,
    total: 1000
  }
}

const spaceShare = {
  id: '1',
  shareType: ShareTypes.space.value,
  collaborator: {
    onPremisesSamAccountName: 'Alice',
    displayName: 'alice'
  },
  role: {
    name: spaceRoleManager.name
  }
}

const formDateFromJSDate = jest.fn().mockImplementation(() => 'ABSOLUTE_TIME')
const formDateFromHTTP = jest.fn().mockImplementation(() => 'ABSOLUTE_TIME')
const refreshShareDetailsTree = jest.fn()

beforeEach(() => {
  formDateFromJSDate.mockClear()
  formDateFromHTTP.mockClear()
  refreshShareDetailsTree.mockReset()
})

describe('Details SideBar Panel', () => {
  it('displays the details side panel', () => {
    const wrapper = createWrapper(spaceMock)
    expect(wrapper).toMatchSnapshot()
  })
})

function createWrapper(spaceResource) {
  const component = {
    ...SpaceDetails,
    setup: () => ({
      spaceImage: '',
      owners: [],
      loadImageTask: {
        isRunning: false,
        perform: jest.fn()
      },
      loadOwnersTask: {
        isRunning: false,
        perform: jest.fn()
      }
    })
  }
  return shallowMount(component, {
    store: new Vuex.Store({
      getters: {
        user: function () {
          return { id: 'marie' }
        }
      },
      modules: {
        Files: {
          namespaced: true,
          state: {
            sharesTree: {}
          },
          getters: {
            highlightedFile: function () {
              return spaceResource
            },
            currentFileOutgoingCollaborators: () => [spaceShare],
            currentFileOutgoingLinks: () => []
          }
        }
      }
    }),
    localVue,
    stubs: stubs,
    directives: {
      OcTooltip
    },
    mixins: [
      {
        methods: {
          formDateFromJSDate,
          formDateFromHTTP
        }
      }
    ],
    provide: {
      displayedItem: {
        value: spaceResource
      }
    }
  })
}
