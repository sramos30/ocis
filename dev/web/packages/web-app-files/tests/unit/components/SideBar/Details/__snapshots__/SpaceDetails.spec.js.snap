// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details SideBar Panel displays the details side panel 1`] = `
<div id="oc-space-details-sidebar">
  <div class="oc-space-details-sidebar-image oc-text-center">
    <oc-icon-stub name="layout-grid" size="xxlarge" class="space-default-image oc-px-m oc-py-m"></oc-icon-stub>
  </div>
  <!---->
  <div>
    <table aria-label="Overview of the information about the selected space" class="details-table">
      <tr>
        <th scope="col" class="oc-pr-s">Last activity</th>
        <td>Invalid DateTime</td>
      </tr>
      <!---->
      <tr>
        <th scope="col" class="oc-pr-s">Manager</th>
        <td><span>alice</span></td>
      </tr>
      <tr>
        <th scope="col" class="oc-pr-s">Quota</th>
        <td>
          <space-quota-stub spacequota="[object Object]"></space-quota-stub>
        </td>
      </tr>
    </table>
  </div>
</div>
`;
